treeherder/model/management/commands/cycle_data.py

```diff
@@ -73,28 +84,192 @@ def prune(id_name, model):
 
 
 class PerfherderCycler(DataCycler):
-    source = PERFHERDER.title()
-    max_runtime = datetime.timedelta(hours=23)
-
-    def __init__(self, days, chunk_size, sleep_time, is_debug=None, logger=None, **kwargs):
-        super().__init__(days, chunk_size, sleep_time, is_debug, logger)
-        if (
-            days < MINIMUM_PERFHERDER_EXPIRE_INTERVAL
-            and settings.SITE_HOSTNAME != 'treeherder-prototype2.herokuapp.com'
-        ):
-            raise ValueError(
-                'Cannot remove performance data that is more recent than {} days'.format(
-                    MINIMUM_PERFHERDER_EXPIRE_INTERVAL
-                )
-            )
+    DEFAULT_MAX_RUNTIME = timedelta(hours=23)
+
+    def __init__(
+        self,
+        chunk_size: int,
+        sleep_time: int,
+        is_debug: bool = None,
+        max_runtime: timedelta = None,
+        strategies: List[RemovalStrategy] = None,
+        **kwargs,
+    ):
+        super().__init__(chunk_size, sleep_time, is_debug)
+        self.started_at = None
+        self.max_runtime = max_runtime or PerfherderCycler.DEFAULT_MAX_RUNTIME
+        self.strategies = strategies or RemovalStrategy.fabricate_all_strategies(chunk_size)
 
     def cycle(self):
-        started_at = datetime.datetime.now()
+        """
+        Delete data older than cycle_interval, splitting the target data
+        into chunks of chunk_size size.
+        """
+        logger.warning(f"Cycling {PERFHERDER.title()} data...")
+        self.started_at = datetime.now()
 
-        PerformanceDatum.objects.cycle_data(
-            self.cycle_interval, self.chunk_size, self.logger, started_at, self.max_runtime
+        try:
+            for strategy in self.strategies:
+                try:
+                    self._delete_in_chunks(strategy)
+                except NoDataCyclingAtAll as ex:
+                    logger.warning('Exception: {}'.format(ex))
+
+            # also remove any signatures which are (no longer) associated with
+            # a job
+            logger.warning('Removing performance signatures with missing jobs...')
+            for signature in PerformanceSignature.objects.all():
+                self._maybe_quit()
+
+                if not PerformanceDatum.objects.filter(
+                    repository_id=signature.repository_id,  # leverages (repository, signature) compound index
+                    signature_id=signature.id,
+                ).exists():
+                    signature.delete()
+        except MaxRuntimeExceeded as ex:
+            logger.warning(ex)
+
+    def _maybe_quit(self):
```

Comment:
I would be more specific here, like: `_quite_runtime_exceeded`. `maybe` is pretty vague and gives room for interpretation/assumption.